### DISCUSSION TITLE
Whose bug is this? Svelte? PB? Me? (Misuse of the Pocketbase SDK?)

### DISCUSSION THREAD

**Discussion by ghost** [pocketbase/pocketbase#1241] [Q&A] [answered]
I know the title might sound a little misleading. I've been using Pocketbase for a while now and on a new app I'm making I noticed there's a bug that has been happening to me in previous projects, this time it's happening more often tho and I cannot seem to pin this down, is this because I'm using the PocketBase SDK wrong in my svelte projects?

## The main problem

Everytime I "start" this svelte application, I do several PocketBase queries using the SDK, namely in three separate components all rendered on the "first page".

Those three components have this structure roughly:

```html
<script lang="ts">
    import { onMount } from "svelte";
    import PocketBase from "pocketbase";
    
    // Config import
    import { fapi_url, api_user, api_user_pass } from "../config";
    
    // Variables
    const pb: PocketBase = new PocketBase(fapi_url);
    let services: any = [];
    onMount(async () => {
        pb.collection('users').authWithPassword(api_user, api_user_pass);
        services = await pb.collection('services').getFullList();
        pb.authStore.clear();
    });
</script>

<h1>{services[0].name}</h1>
```

However this throws a: `Services.svelte:30 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'name')`

So I try the first approach, define a `Service.d.ts` file:

```typescript
export interface Service {
  id: string,
  collectionId: string,
  collectionName: string,
  created: string,
  updated: string,
  name: string,
  image: string,
  available: number,
  description: string,
}
```

Everything looks file I should just change my array definition:

```typescript
// -- Snip
    const pb: PocketBase = new PocketBase(fapi_url);
    let services: Product[] = [];
    onMount(async () => {
        pb.collection('users').authWithPassword(api_user, api_user_pass);
        services = await pb.collection('services').getFullList();
        pb.authStore.clear();
    });
// -- Snip
```

But this causes a HMR error due to the same "undefined" properties, now the whole page is broken and doesn't load at all.

So I resort to a four-letter word and realize that, I only have 8 services inside my PocketBase instance, so I might as well mess things up more:

```typescript
    const pb: PocketBase = new PocketBase(fapi_url);
    let services: any = [{}, {}, {}, {}, {}, {}, {}, {}, {}];
    onMount(async () => {
        pb.collection('users').authWithPassword(api_user, api_user_pass);
        services = await pb.collection('services').getFullList();
        pb.authStore.clear();
    });
```

Yes I know it's awful. But this produces a kind of unexpected result, you see. It sometimes loads the services properly and sometimes it doesn't:

When it works I do get those services loaded:
![image](https://user-images.githubusercontent.com/24773698/207002126-206b8a8c-40c0-4861-a0cd-86dad0a5692d.png)

When it doesnt I get an undefined:

![image](https://user-images.githubusercontent.com/24773698/207001534-3cd7b4e5-f355-431a-9cf5-3ebf9cc46126.png)

This happens randomly between reloads, if you do a <kbd>Ctrl + F5</kbd> reload all of the page data is processed properly, but when doing normal <kbd>F5</kbd> reloads, the data might or might not appear, seemingly "when it wants to".

I'm still at doubt if it's because my `onMount();` function has an async function inside it:

```typescript
    onMount(async () => {
        pb.collection('users').authWithPassword(api_user, api_user_pass);
        services = await pb.collection('services').getFullList();
        pb.authStore.clear();
    });
```

I've read that this might cause some problems in [this issue](https://github.com/sveltejs/svelte/issues/4927) and I don't know if I'm misusing the PocketBase SDK and going directly into that svelte pitfall. 

I've also tried using `{#await onMonut}` blocks but I get a compiler warning that reads: `await has no effect on the type of this expression` so I guess that isn't the correct approach.

Any ideas on this matter? :thinking:

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand the described issue. I recommend adding a check before accessing your `services` items directly, see for example https://svelte.dev/tutorial/onmount.

Additionally, I noticed that you are missing `await` in front of the `authWithPassword` call.

**Reply by ghost** [reply]
I've tried adding the `await` keyword to the authWithPassword. It did get rid of some errors in a lot of components.

I'm re-reading the JS-SDK instructions and I'm wondering if changing the array assingments from `array = await pb.collection(...)...` to `pb.collection(...).then((x) => {...})` will make any improvements.

Idk if they are the same as far as I know `async/await` is a readability thing unless I'm missing something here.

Thank you for your response :)

**Reply by ganigeorgiev** [reply]
>  I'm wondering if changing the array assingments from array = await pb.collection(...)... to pb.collection(...).then((x) => {...}) will make any improvements.

They serve the same purpose, but please note that when using `await` it will block the execution of any code after the statement within the defined async scope while the `Promise` callback syntax doesn't.

I don't think this is a Svelte, nor a SDK issue. I haven't tested it but I'm not sure that making onMount async will actually always block rendering the component and this is why in the linked example https://svelte.dev/tutorial/onmount they are showing a "Loading..." text before accessing the fetched items.

**Reply by ghost** [reply]
My other theory is that I'm using the SDK in a wrong way on another component and that's blocking the main thread or it waits until the previous promise(s) just timeout and therefore I get empty arrays.

**Reply by ganigeorgiev** [reply]
You can try adding a try/catch block and check if the methods throws any error. It could also be an auto cancellation issue - https://github.com/pocketbase/js-sdk#auto-cancellation.
But once again, I don't think it is safe to directly access the array index before the items are loaded since the onMount doesn't block the rendering (at least not on client-side - https://svelte.dev/repl/3bdc4c452dcb43c9b8c141f44eebabff?version=3.54.0).

**Reply by ghost** [reply]
I managed to solve the errors in the console by properly calling `await` on all of the neccesary functions. However it seems that this error goes far beyond the console errors, as the "undefined" fields take too long to be "filled in" with information (Around 4-5 seconds)

Is the getFullList() method heavy on the main thread? :thinking: I render at least 4 components that make use of this method (with a reduced batch-size).

I think it's the authentication part that is slowing down the main thread as each component has those "three" steps on it's `onMount()` method, the
authentication, the `getFullList()` and finally the `authStore.clear()` method call.

**Reply by ganigeorgiev** [reply]
> Is the getFullList() method heavy on the main thread?

It depends on the number of elements the collection have and the underlying query.
Authentication is slow but not 5 seconds slow. Something else is blocking your rendering.

But in any case, I don't recommend authenticating in each individual component. Usually it should happen once on page load and then reuse the state (hence the `AuthStore` name).

**Reply by ghost** [reply]
I think I should search how to do that single auth.

My items are between 20-60 items per "load". In some cases and it does take time. I've profiled the application but the total graph shows around 4-5 seconds of "idle" time. 

I know it's not autocancellation because I don't have any errors in the console, but after those 4-5 seconds I get a huge "authentication" messages "spam":

![image](https://user-images.githubusercontent.com/24773698/207242909-cc3adfcb-25d6-4040-bf90-543f9c7328b5.png)

**Reply by ganigeorgiev** [reply]
@VentGrey You can get a better idea when and how much a request has taken by inspecting the "Network" tab and filter the fetch/xhr requests (if they are executed in the browser).

It could also be just a slow server response due to uneficient filter query. You can also start the PocketBase executable with `--debug` to see the executed sql statements.

**Reply by ghost** [reply]
I'm exploring the network tab and seeing that indeed, what you have told me is true, the authentication-per-component method is doing strange things.

The authentication request is being sent twice (idk why) and it indeed takes a long time:

(Sorry for the 4K - HD image)
![image](https://user-images.githubusercontent.com/24773698/207263855-88b83679-e639-400c-adcb-0c93385252ab.png)

Both requests with the same token and the same data inside them :thinking:

**Reply by ganigeorgiev** [reply]
The 204 are probably `OPTIONS`/preflight requests and this is ok.

I'm not sure how to help. The above screenshot doesn't provide much information how the requests are performing and what is being sent.

Where your code is executed - client-side, in a server-side node handler or both?
Are the response times different when you perform the requests with curl or rest client (eg. postman, insomnia, etc.)?
Could you reproduce the issue locally or only on prod (maybe there is a firewall/proxy/vpn/etc.)?

**Reply by ghost** [reply]
Everything is client-side using a normal svelte-vite project. :thinking: I've been trying to take the `.authWithPassword()` method to a more "global" level.

**Reply by ganigeorgiev** [reply]
@VentGrey If your app is entirely client-side (eg. SPA) and there is no SSR involved, then it is safe to move your PocketBase initialization in its own file, and import it in your individual components.

_Note:_ If your app is only client-side then the way you are loading the `api_user` and `api_user_pass` is not correct since everyone will have access to your config. Make sure that your code is running where you expect to avoid leaking sensitive information. If you are still experiencing issues please try making direct API requests using curl, postman, insomnia, etc. to compare the loading time and to rule out a PocketBase issue. You can also upgrade to the [latest v0.10.1](https://github.com/pocketbase/pocketbase/releases) (it comes with some performance improvements).

**Reply by ghost** [reply]
I ended up doing the following, tho I don't know if this was the best approach.

1- Setup a pocketbase non-admin user for read-only purposes.
2- Use Vite to import those values from a .env file.
3- Upgrade to 0.10.1 (The "undefined" times did improve from 4s to ~1s)
4- Stop relying on `onMout()` and make my own wrapper.
5- Rate limit the requests/s in NGINX
6- Authenticate at `App.svelte` level, only dropping the auth store when `App.svelte` is destroyed.

As no sensitive information is really being held there there is not much problem in reality. In case someone else faces this issue I'll provide code in the comments, also, the code for which I asked help is in a public repo and is fully FLOSS so :sweat_smile: idk what's best for everyone.

## In App.svelte

```typescript
    // Database imports
    import PocketBase from 'pocketbase'; 
    
    // Database usage
    const pb: PocketBase = new PocketBase(fapi_url);
    onMount(async () => { await pb.collection('users').authWithPassword(api_user, api_user_pass); });

    onDestroy(() => { pb.authStore.clear(); });
```

## The PocketBase wrap function at pbwrap.ts

```typescript
import PocketBase from "pocketbase";
import type { Record } from "pocketbase";

/**
 * Wraps the `getFullList` method to reuse on other components.
 *
 * @export
 * @param {PocketBase} pb
 * @param {string} table
 * @return {*}  {Promise<Record[]>}
 */
export async function getTables(pb: PocketBase, table: string): Promise<Record[]> {
    return await pb.collection(table).getFullList();
}
```
### And an example usage inside a component:

```typescript
    import PocketBase from 'pocketbase';
    import type { Record } from 'pocketbase';
    const pb: PocketBase = new PocketBase(fapi_url);
    const records: Promise<Record[]> = getTables(pb, 'products');
```
Then you can iterate normally after awaiting the `records` variable:

```html
<section class="container">
    {#await records} <p>Loading products...</p>{:then data}     
    {#each data as record}
        <h1>{record.name}</h1> <!-- Fields may vary depending on your data -->
    {/each}
    {/await}
</section>
```


Again @ganigeorgiev thanks a lot for your time and support! :pray:
